src/server/graphics/platform_probe.cpp

```diff
@@ -231,5 +300,238 @@ auto mir::graphics::rendering_modules_for_device(
         {
             return probe_rendering_module(platforms, module, options, console);
         },
-        modules);
+        modules,
+        TypePreference::prefer_hardware);
+}
+
+namespace
+{
+auto split_on(std::string const& tokens, char delimiter) -> std::vector<std::string>
+{
+    std::string token;
+    std::istringstream token_stream{tokens};
+
+    std::vector<std::string> result;
+    while (std::getline(token_stream, token, delimiter))
+    {
+        result.push_back(token);
+    }
+
+    return result;
+}
+
+auto select_platforms_from_list(std::string const& selection, std::vector<std::shared_ptr<mir::SharedLibrary>> const& modules)
+    -> std::vector<std::shared_ptr<mir::SharedLibrary>>
+{
+    std::vector<std::shared_ptr<mir::SharedLibrary>> selected_modules;
+    std::vector<std::string> found_module_names;
+
+    // Our platform modules are a comma-delimited list.
+    auto requested_modules = split_on(selection, ',');
+
+    for (auto const& module : modules)
+    {
+        try
+        {
+            auto describe_module = module->load_function<mg::DescribeModule>(
+                "describe_graphics_module",
+                MIR_SERVER_GRAPHICS_PLATFORM_VERSION);
+            auto const description = describe_module();
+            found_module_names.emplace_back(description->name);
+
+            if (auto const i = std::find(requested_modules.begin(), requested_modules.end(), description->name);
+                i != requested_modules.end())
+            {
+                selected_modules.push_back(module);
+                requested_modules.erase(i);
+            }
+        }
+        catch (std::exception const&)
+        {
+            // Should we log anything here?
```

Comment:
Should probably resolve that question before merging…

Reply:
Hm. Maybe the resolution should be to remove the `try` entirely, and make “we're passed a list of graphics platform modules” a precondition (one that already hold).